### PR TITLE
tool init

### PR DESCRIPTION
**Pull request by godrei**

### CODE REVIEW

**Comment by godrei** [comment]
Some usage instruction would have been great.

Based on your `main.go` file, where you have a relative import path (`"bitrise-app-analyser/analyser"`), I thought that the project uses [go modules](https://github.com/golang/go/wiki/Modules#quick-start), but the necessary `go.mod` file was missing.

Otherwise based on my best knowledge you have to use absolute import path: https://golang.org/doc/code.html#ImportPaths

I pushed your source code to this repository ( to make it easy to share my review) and fixed the import path (`import /bitrise-app-analyser/analyser"` -> `import "github.com/godrei/bitrise-app-analyser/analyser"`) in a [subsequent commit](
https://github.com/godrei/bitrise-app-analyser/pull/1/commits/34302a2e828c5e254f1bdb20291032fccd3100c7).

To test this project:

-  run `git clone https://github.com/godrei/bitrise-app-analyser.git -b development $GOPATH/src/github.com/godrei/bitrise-app-analyser`
- `cd $GOPATH/src/github.com/godrei/bitrise-app-analyser`
- `go run main.go <some/ipa/path>`

**Comment by godrei** [comment]
The first test against my test IPA file crashed:

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/godrei/bitrise-app-analyser/analyser.ipaAnalyser.getAppInfo(0xc0000a54a0, 0x1)
	/Users/godrei/Develop/go/src/github.com/godrei/bitrise-app-analyser/analyser/ipaAnalyser.go:45 +0x435
github.com/godrei/bitrise-app-analyser/analyser.ipaAnalyser.PrintAppInfo(0xc0000a54a0)
	/Users/godrei/Develop/go/src/github.com/godrei/bitrise-app-analyser/analyser/ipaAnalyser.go:19 +0x40
main.handleInputFile(0xc0000503d0)
	/Users/godrei/Develop/go/src/github.com/godrei/bitrise-app-analyser/main.go:35 +0xf4
main.main()
	/Users/godrei/Develop/go/src/github.com/godrei/bitrise-app-analyser/main.go:14 +0x72
exit status 2
```

The embedded Info.plist file can be an `Apple binary property list` inside the .ipa file, this was the case with my test file. It seems the Info.plist format [can be changed in Xcode](https://stackoverflow.com/a/14370326) (in recent Xcode version this is the default format).

The file type can be checked by unzipping the IPA file and calling: 
```
$ file Payload/Runner.app/Info.plist

> Payload/Runner.app/Info.plist: Apple binary property list
```

Using `plutil` you can convert the binary plist into an `XML 1.0 document text, ASCII text` by calling `plutil -convert xml1 Payload/Runner.app/Info.plist`

The binary property list is not working well with your [regex patterns](https://github.com/godrei/bitrise-app-analyser/pull/1/files#diff-aa70c7ea7064f78df2887bc8ec3c48d3R38), this caused the crash.

__For an IPA file, which has a non-binary Info.plist embedded, the tool works well.__

**Comment by godrei** [comment]
My second test using an APK file also failed:

```
2019/01/17 16:23:29 fork/exec /Users/godrei/Library/Android/sdk/build-tools/22.0.1
28.0.2
28.0.3/aapt: no such file or directory
exit status 1
```

This is cased by an incorrect `aapt` command call. I have multiple build tools version installed, so `getBuildToolsVersion` returns a list of my installed build tools versions:

```
22.0.1
28.0.2
28.0.3
```

This way the build aapt command call looks like: 

```
/Users/godrei/Library/Android/sdk/build-tools/22.0.1
28.0.2
28.0.3/aapt dump badging <APK_PATH>
```

__If i have only one build tool version, the tool works well for an APK file.__

File: analyser/analyser.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+//detectPlatform returns one of the platform identifier string constants or an empty string
+func detectPlatform(fileName *string) string {
+	if strings.EqualFold(*fileName, "AndroidManifest.xml") {
+		//fmt.Println("Android app recognized")
+		return platformAndroid
+	}
+
+	plistPathRegex, err := regexp.Compile(`^Payload/[^/]+/Info\.plist$`)
+	if err != nil {
+		log.Fatal(err)
```
Comment (godrei) [inline analyser/analyser.go:57]: This inner function should not fail, as it makes hard to follow, where the code can exist.
Most go function (which can produce an abnormal state) returns an error and lets the caller/client handle it.
It is a good practice to exit only in the main function.

File: analyser/analyser.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+//AppInfo is a struct, holding the recognisable information of the mobile app package
+type AppInfo struct {
+	AppId    string
+	Version  string
+	Build    string
+	IconPath string
+}
+
+//GetAnalyser returns the appropriate analyser implementation, based on the contents of the package
+func GetAnalyser(rc *zip.ReadCloser, originalFilePath *string) (IAnalyser, error) {
```
Comment (godrei) [inline analyser/analyser.go:31]: it would be slightly better if this function would only receive the originalFilePath, as the zip read closer only used in this function, but originalFilePath is a required input for this function.

File: analyser/apkAnalyser.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+//apkAnalyser is an IAnalyser implementation for Android platform
+type apkAnalyser struct {
+	filePath *string
+}
+
+//PrintAppInfo fetches and prints information about the Android app package
+func (analyser apkAnalyser) PrintAppInfo() {
+	appInfo := analyser.getAppInfo()
+	fmt.Printf("Recognised app info:\n "+
+		"\tPackage Name: %+v\n "+
```
Comment (godrei) [inline analyser/apkAnalyser.go:21]: `%s` would be enough as every field of appInfo is a string.  `%v`: the value in a default format when printing structs, the plus flag (`%+v`) adds field names: https://golang.org/pkg/fmt/#hdr-Printing

File: analyser/apkAnalyser.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+	return &AppInfo{packageName, versionName, versionCode, appIcon}
+}
+
+//getBuildToolsVersion returns the first available build tools version
+//available in the current Android SDK installation
+func getBuildToolsVersion() string {
+	cmd := "ls"
+	param := os.ExpandEnv("$ANDROID_HOME") + "/build-tools"
+	output, err := exec.Command(cmd, param).CombinedOutput()
```
Comment (godrei) [inline analyser/apkAnalyser.go:70]: in case of multiple build tools versions, the output is a multi-line string, which will lead to an error later

File: analyser/apkAnalyser.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		appInfo.Version,
+		appInfo.Build,
+		appInfo.IconPath)
+}
+
+//getAppInfo returns populated AppInfo
+func (analyser apkAnalyser) getAppInfo() *AppInfo {
+	buildToolsVersion := getBuildToolsVersion()
+	buildToolPath := "/build-tools/" + buildToolsVersion
+	androidSDKPath := os.ExpandEnv("$ANDROID_HOME")
```
Comment (godrei) [inline analyser/apkAnalyser.go:35]: `ANDROID_HOME` is also expanded in `getBuildToolsVersion`, getBuildToolsVersion could return the full build tool path, this way, you could remove this line

File: analyser/apkAnalyser.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	if len(androidSDKPath) == 0 {
+		log.Fatal("$ANDROID_HOME environment variable not set")
+		return nil
+	}
+	cmd := androidSDKPath + buildToolPath + "/aapt"
```
Comment (godrei) [inline analyser/apkAnalyser.go:40]: `filepath.Join()` is a secure way to build file path in go

File: analyser/apkAnalyser.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	appIdRegex, err := regexp.Compile(`package: name='(?P<package>[^']+)'`)
+	versionCodeRegex, err := regexp.Compile(`versionCode='(?P<code>[^']+)'`)
+	versionNameRegex, err := regexp.Compile(`versionName='(?P<name>[^']+)'`)
+	appIconRegex, err := regexp.Compile(`icon='(?P<name>[^']+)'`)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	packageName := string(appIdRegex.FindSubmatch(output)[1])
```
Comment (godrei) [inline analyser/apkAnalyser.go:57]: if the regex would not match this line would lead to a crash (`index out of range`)

File: analyser/ipaAnalyser.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+//readFile returns the ReadCloser's content as a string
+func readFile(rc io.ReadCloser) string {
+	var result string
+	buf := new(bytes.Buffer)
+	_, err := buf.ReadFrom(rc)
+	result += string(buf.Bytes())
+	err = rc.Close()
+
+	if err != nil {
```
Comment (godrei) [inline analyser/ipaAnalyser.go:63]: this way the `buf.ReadFrom(rc)` function's error can be swallowed, if `err = rc.Close()` also throws an error

**Review by godrei — COMMENTED** [review]
Please have a look at my notes.

In general:

Thanks for working on this tool! I really appropriate, that you solved both the ipa and apk analyser part of the task and that you have write tests ❤️! 

Your code faced some edge cases on my dev environment, but i really like that you used an interface for the analyzers.

You understand the task and the general solution is a good solution for this (we also use `aapt` to analyse an apk, in case of ipa we parse the plist as an xml and we apply a model on it).
